testnet/stacks-node/src/nakamoto_node/stackerdb_listener.rs

```diff
@@ -0,0 +1,444 @@
+                                "signature" => %signature,
+                                "block_signer_signature_hash" => %block_sighash,
+                                "slot_id" => slot_id,
+                            );
+                            continue;
+                        }
+
+                        if Self::fault_injection_ignore_signatures() {
+                            warn!("StackerDBListener: fault injection: ignoring well-formed signature for block";
+                                "block_signer_sighash" => %block_sighash,
+                                "signer_pubkey" => signer_pubkey.to_hex(),
+                                "signer_slot_id" => slot_id,
+                                "signature" => %signature,
+                                "signer_weight" => signer_entry.weight,
+                                "total_weight_signed" => block.total_weight_signed,
+                            );
+                            continue;
+                        }
+
+                        if !block.gathered_signatures.contains_key(&slot_id) {
+                            block.total_weight_signed = block
+                                .total_weight_signed
+                                .checked_add(signer_entry.weight)
+                                .expect("FATAL: total weight signed exceeds u32::MAX");
+                        }
+
+                        info!("StackerDBListener: Signature Added to block";
+                            "block_signer_sighash" => %block_sighash,
+                            "signer_pubkey" => signer_pubkey.to_hex(),
+                            "signer_slot_id" => slot_id,
+                            "signature" => %signature,
+                            "signer_weight" => signer_entry.weight,
+                            "total_weight_signed" => block.total_weight_signed,
+                            "tenure_extend_timestamp" => tenure_extend_timestamp,
+                            "server_version" => metadata.server_version,
+                        );
+                        block.gathered_signatures.insert(slot_id, signature);
+                        block.responded_signers.insert(signer_pubkey);
+
+                        if block.total_weight_signed >= self.weight_threshold {
+                            // Signal to anyone waiting on this block that we have enough signatures
+                            cvar.notify_all();
+                        }
+
+                        // Update the idle timestamp for this signer
+                        self.update_idle_timestamp(
+                            signer_pubkey,
+                            tenure_extend_timestamp,
+                            signer_entry.weight,
+                        );
+                    }
+                    SignerMessageV0::BlockResponse(BlockResponse::Rejected(rejected_data)) => {
+                        let (lock, cvar) = &*self.blocks;
```

Comment:
Same here -- this needlessly marries the synchronization implementation (something that can be hidden) to the business logic.

Reply:
Improved in `StackerDBListenerComms`.